ocaml/typing/ctype.ml

```diff
@@ -2221,14 +2221,12 @@ let () =
   Env.constrain_type_jkind := constrain_type_jkind
 
 let check_type_externality env ty ext =
-  let check_sub ty_jkind =
-    Jkind.(Externality.le (get_externality_upper_bound ty_jkind) ext)
+  let upper_bound =
+    Jkind.set_externality_upper_bound (Jkind.any ~why:Dummy_jkind) ext
```

Comment:
This new version will be a little slower, because it has a bunch more checks to make. It probably doesn’t matter, and am happy for you to resolve if you think the benefit of this patch is worth the runtime cost.

Reply:
On the flip side, the old implementation had to make a closure in the `constrain_type_jkind` case, and now there’s no allocation there. My guess is that we actually care about externality more often than about kinds in general, but I could be wrong there.

Reply:
I agree there is some cost here. The check is pretty cheap even in the new world, though, and some refactoring here is necessary to support product layouts sensibly.

This PR is an intermediate step - I think when we discussed this a few weeks ago we concluded the cleanest thing may be for `constrain_type_jkind` to handle the modal upper bounds separately from the layout, which may well end up exposing a cheaper way to do this check. I'll leave a CR here to remember to revisit after product layouts.

Reply:
Yes, OK.